src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,102 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    static final int ARRAY_MIN_SIZE = 10;
+
+    private T[] array = (T[]) new Object[ARRAY_MIN_SIZE];
+    private int size;
+    private int newCapacity;
+    private T[] arrayGrow;
+
     @Override
     public void add(T value) {
-
+        if (size >= array.length) {
+            grow(array);
+            array = arrayGrow;
+        }
+        array[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index >= 0 && index <= size) {
+            if (size >= array.length) {
+                grow(array);
+                array = arrayGrow;
+            }
+            System.arraycopy(array, index, array, index + 1, size - index);
+            array[index] = value;
+            size++;
+        } else {
+            throw new ArrayListIndexOutOfBoundsException("Invalid index: " + index);
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        isIndexValid(index);
+        return array[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        isIndexValid(index);
+        array[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (isIndexValid(index)) {
+            T removeElement = array[index];
+            System.arraycopy(array, index + 1, array, index, size - index - 1);
+            size--;
+            return removeElement;
+        }
+        throw new ArrayListIndexOutOfBoundsException("Invalid index: " + index);
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        for (int i = 0; i < size; i++) {
+            if (element == array[i] || array[i] != null && array[i].equals(element)) {
+                T removeElement = array[i];
+                remove(i);
+                return removeElement;
+            }
+        }
+        throw new NoSuchElementException("Element: " + element + " is missing");
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void grow(T[] array) {
+        newCapacity = array.length + (array.length >> 1);
+        arrayGrow = (T[]) new Object[newCapacity];
+        System.arraycopy(array, 0, arrayGrow, 0, size);
+    }
```

Comment:
We also should not receive `array` as a parameter, it should be just as `void grow()`
you have everything you need to do this, just call `array` directly from this method and do not add redundant one-time use fields to the class
```suggestion
    private void grow() {
        int newCapacity = array.length + (array.length >> 1);
        T[] temp = (T[]) new Object[newCapacity];
        System.arraycopy(array, 0, temp, 0, size);
        array = temp;
    }
```